File: MouseInvasion/app/src/main/java/br/odb/mouseinvasion/BeamCheese.java
/**
 * 
 */
package br.odb.mouseinvasion;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Paint.Style;

/**
 * @author monty
 * 
 */
public class BeamCheese extends BonusCheese {

	private int kind;
	private static Bitmap vertical;
	private static Bitmap horizontal;
	private static Bitmap multi;

	/**
	 * 
	 */
	public BeamCheese(int kind) {
		super();

		if (multi == null)
			multi = BitmapFactory.decodeResource(CheeseDefenderActivity
					.getInstance().getResources(), R.drawable.multi);

		if (vertical == null)
			vertical = BitmapFactory.decodeResource(CheeseDefenderActivity
					.getInstance().getResources(), R.drawable.vertical);

		if (horizontal == null)
			horizontal = BitmapFactory.decodeResource(CheeseDefenderActivity
					.getInstance().getResources(), R.drawable.horizontal);

		this.kind = kind;

		switch (kind) {
		case 3:
			super.frames[0] = horizontal;
			break;
		case 4:
			super.frames[0] = vertical;
			break;
		case 5:
			super.frames[0] = multi;
			break;
		}
	}

	@Override
	public boolean isHit(GameObject go) {
		boolean beamHit;

		if (isExploding()) {
			switch (kind) {
			case 3:
				beamHit = ((go.position.y >= (position.y - 10)) && (go.position.y <= (position.y + 10)));
				break;
			case 4:
				beamHit = ((go.position.x >= (position.x - 10)) && (go.position.x <= (position.x + 10)));
				break;
			default:
				beamHit = ((go.position.y >= (position.y - 10)) && (go.position.y <= (position.y + 10)));
				beamHit = beamHit
						|| (((go.position.x >= (position.x - 10)) && (go.position.x <= (position.x + 10))));
				break;
			}
		} else {
			beamHit = false;
		}
		return super.isHit(go) || beamHit;
	}

	@Override
	public void draw(Canvas canvas, Paint paint) {

		super.draw(canvas, paint);

		if (super.getExplosionTime() > 0) {
			paint.setStyle(Style.FILL_AND_STROKE);
			paint.setARGB((int) ((255 * getExplosionTime()) / 2000), 255, 255,
					0);

			if (kind == 3)
				canvas.drawLine(0, getPosition().y, CheeseDefenderView
						.getInstance().getWidth(), getPosition().y, paint);

			if (kind == 4)
				canvas.drawLine(getPosition().x, 0, getPosition().x,
						CheeseDefenderView.getInstance().getHeight(), paint);

			if (kind == 5) {
				canvas.drawLine(getPosition().x, 0, getPosition().x,
						CheeseDefenderView.getInstance().getHeight(), paint);
				canvas.drawLine(0, getPosition().y, CheeseDefenderView
						.getInstance().getWidth(), getPosition().y, paint);

			}
		}
	}
}

File: MouseInvasion/app/src/main/java/br/odb/mouseinvasion/BeamCheese.kt
package br.odb.mouseinvasion

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Paint

open class BeamCheese(private val kind: Int,
                      private val screenWidth : Float,
                      private val screenHeight : Float,
                      context : Context,
                      soundEnabled : Boolean) : BonusCheese(context, soundEnabled) {

    companion object {
        private var vertical: Bitmap? = null
        private var horizontal: Bitmap? = null
        private var multi: Bitmap? = null
    }

    init {
        if (multi == null) multi = BitmapFactory.decodeResource(
            context.resources, R.drawable.multi
        )
        if (vertical == null) vertical = BitmapFactory.decodeResource(
            context.resources, R.drawable.vertical
        )
        if (horizontal == null) horizontal = BitmapFactory.decodeResource(
            context.resources, R.drawable.horizontal
        )

        when (kind) {
            3 -> super.frames[0] = horizontal
            4 -> super.frames[0] = vertical
            5 -> super.frames[0] = multi
        }
    }

    override fun isHit(go: GameObject): Boolean {

        var beamHit: Boolean

        if (isExploding) {
            when (kind) {
                3 -> beamHit = go.position.y >= position.y - 10 && go.position.y <= position.y + 10
                4 -> beamHit = go.position.x >= position.x - 10 && go.position.x <= position.x + 10
                else -> {
                    beamHit = go.position.y >= position.y - 10 && go.position.y <= position.y + 10
                    beamHit = (beamHit
                            || go.position.x >= position.x - 10 && go.position.x <= position.x + 10)
                }
            }
        } else {
            beamHit = false
        }

        return super.isHit(go) || beamHit
    }

    override fun draw(canvas: Canvas, paint: Paint) {

        super.draw(canvas, paint)

        if (explosionTime > 0) {
            paint.style = Paint.Style.FILL_AND_STROKE
            paint.setARGB(
                (255 * explosionTime / 2000).toInt(), 255, 255,
                0
            )
            if (kind == 3) canvas.drawLine(
                0f, position.y.toFloat(), screenWidth, position.y.toFloat(), paint
            )
            if (kind == 4) canvas.drawLine(
                position.x.toFloat(), 0f, position.x.toFloat(), screenHeight, paint
            )
            if (kind == 5) {
                canvas.drawLine(
                    position.x.toFloat(), 0f, position.x.toFloat(), screenHeight, paint
                )
                canvas.drawLine(
                    0f, position.y.toFloat(), screenWidth, position.y.toFloat(), paint
                )
            }
        }
    }
}